kymatio/scattering1d/frontend/base_frontend.py

```diff
@@ -368,4 +523,1496 @@ def _document(cls):
             n=cls._doc_array_n)
 
 
-__all__ = ['ScatteringBase1D']
+class TimeFrequencyScatteringBase1D():
+    def __init__(self, J_fr=None, Q_fr=2, F=None, implementation=None,
```

Comment:
it's not clear to me what `implementation=None` does? why not `implementation=1` by default?

Reply:
implementation `1` seems like the safest default option. I guess the `None` option lets the user build their own implementation?

Reply:
`None` will set any params unspecified by the user to that of `implementation=1`.

`not None` will _override_ any params specified by the user by that of `3` (or whatever).

It's the only way to distinguish user-supplied args from defaulted args, besides defaulting everything else to `None` in the function declaration, which we don't want.

Reply:
I propose this:
(beginner) by default, `implementation=1`
(intermediate) the user can switch to a different implementation by setting `implementation=2`, `3`, etc.
(advanced) the user may pass `implementation` as a dictionary containing `average_fr` and friends.

The main benefit of this is that we don't have parameters overriding other parameters under the hood. Every user-defined choice has an effect.

The second benefit is that it allows a learning curve for JTFS. First without `implementation` (implicitly =1), then with integer-valued presets, then with dict-valued custom settings (useful in debugging and research!).

The third benefit is that my proposal is already partially achieved by @OverLordGoldDragon via `self._implementation_args`, so i don't expect there's a lot to change here.

Reply:
That's a nice way to clean up the namespace and support more args without worry for clutter, yes. Some problems:

 1. Currently two very basic parameters, `average_fr` and `out_type`, vary by `implementation`: do we really want to hide these?
 2. How would docs change? Some params have long descriptions and I'd really rather not have them all indented and packed under the grand umbrella `implementation`. Perhaps a note somewhere, "these kwargs are accessible via `implementation`"?

> don't have parameters overriding other parameters under the hood.

Agreed that's a benefit. While it's documented, it'd be safer to throw a warning; problem is, we'd create duplication as we'd have to store defaults both in `def` and elsewhere, to be able to compare.

Reply:
1. yes, frankly i have no problem with hiding `average_fr` and `out_type` behind implementation
2. yes, good idea

From this discussion it seems like we should either put `implementation` to full use (remove other kwargs) or remove `implementation` itself, as proposed by @rastegah. Leaving both will lead to a never-ending stream of silent bugs in user code.

"There should be one-- and preferably only one --obvious way to do it." — Zen of Python


> While it's documented, it'd be safer to throw a warning

Warnings are too easily ignored and should only be used as a last resort. Error messages are almost always preferable.